// sandstone/env.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONFLICT_STRATEGIES = exports.PACK_UID = exports.NAMESPACE = void 0;
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
exports.NAMESPACE = typeof window === 'undefined' ? process.env.NAMESPACE : window.NAMESPACE;
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
exports.PACK_UID = typeof window === 'undefined' ? process.env.PACK_UID : window.PACK_UID;
const { GENERAL_CONFLICT_STRATEGY, ADVANCEMENT_CONFLICT_STRATEGY, LOOT_TABLE_CONFLICT_STRATEGY, MCFUNCTION_CONFLICT_STRATEGY, PREDICATE_CONFLICT_STRATEGY, RECIPE_CONFLICT_STRATEGY, TAG_CONFLICT_STRATEGY, } = process.env;
const DEFAULT_STRATEGY = 'warn';
let GENERAL_STRATEGY = DEFAULT_STRATEGY;
if (GENERAL_CONFLICT_STRATEGY && isBasic(GENERAL_CONFLICT_STRATEGY)) {
    GENERAL_STRATEGY = GENERAL_CONFLICT_STRATEGY;
}
exports.CONFLICT_STRATEGIES = {
    DEFAULT: DEFAULT_STRATEGY,
    GENERAL: GENERAL_STRATEGY,
    ADVANCEMENT: GENERAL_STRATEGY,
    LOOT_TABLE: GENERAL_STRATEGY,
    MCFUNCTION: GENERAL_STRATEGY,
    PREDICATE: GENERAL_STRATEGY,
    RECIPE: GENERAL_STRATEGY,
    TAG: GENERAL_STRATEGY,
};
function isBasic(strategy) {
    return strategy === 'throw' || strategy === 'replace' || strategy === 'warn' || strategy === 'ignore';
}
if (isBasic(ADVANCEMENT_CONFLICT_STRATEGY)) {
    exports.CONFLICT_STRATEGIES.ADVANCEMENT = ADVANCEMENT_CONFLICT_STRATEGY;
}
if (isBasic(LOOT_TABLE_CONFLICT_STRATEGY)) {
    exports.CONFLICT_STRATEGIES.LOOT_TABLE = LOOT_TABLE_CONFLICT_STRATEGY;
}
if ((isBasic(MCFUNCTION_CONFLICT_STRATEGY) || MCFUNCTION_CONFLICT_STRATEGY === 'append' || MCFUNCTION_CONFLICT_STRATEGY === 'prepend')) {
    exports.CONFLICT_STRATEGIES.MCFUNCTION = MCFUNCTION_CONFLICT_STRATEGY;
}
if (isBasic(PREDICATE_CONFLICT_STRATEGY)) {
    exports.CONFLICT_STRATEGIES.PREDICATE = PREDICATE_CONFLICT_STRATEGY;
}
if (isBasic(RECIPE_CONFLICT_STRATEGY)) {
    exports.CONFLICT_STRATEGIES.RECIPE = RECIPE_CONFLICT_STRATEGY;
}
if (isBasic(TAG_CONFLICT_STRATEGY)) {
    exports.CONFLICT_STRATEGIES.TAG = TAG_CONFLICT_STRATEGY;
}
