// sandstone/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Variable = exports.Data = exports.Selector = exports.Objective = exports.SelectorClass = exports.rel = exports.loc = exports.abs = exports._ = exports.Recipe = exports.LootTable = exports.Tag = exports.Predicate = exports.Advancement = exports.MCFunction = exports.sleep = exports.BasePath = exports.savePack = void 0;
const init_1 = require("./init");
// CORE
exports.savePack = init_1.dataPack.save, exports.BasePath = init_1.dataPack.BasePath, exports.sleep = init_1.dataPack.sleep;
_a = init_1.dataPack.basePath, exports.MCFunction = _a.MCFunction, exports.Advancement = _a.Advancement, exports.Predicate = _a.Predicate, exports.Tag = _a.Tag, exports.LootTable = _a.LootTable, exports.Recipe = _a.Recipe;
var init_2 = require("./init");
Object.defineProperty(exports, "_", { enumerable: true, get: function () { return init_2._; } });
// COMMANDS
__exportStar(require("./commandsOnly"), exports);
// VARIABLES
var _1 = require("./index");
Object.defineProperty(exports, "abs", { enumerable: true, get: function () { return _1.absolute; } });
Object.defineProperty(exports, "loc", { enumerable: true, get: function () { return _1.local; } });
Object.defineProperty(exports, "rel", { enumerable: true, get: function () { return _1.relative; } });
__exportStar(require("./variables/Coordinates"), exports);
__exportStar(require("./variables/JSONTextComponentClass"), exports);
__exportStar(require("./variables/NBTs"), exports);
__exportStar(require("./variables/Objective"), exports);
__exportStar(require("./variables/parsers"), exports);
__exportStar(require("./variables/Score"), exports);
var Selector_1 = require("./variables/Selector");
Object.defineProperty(exports, "SelectorClass", { enumerable: true, get: function () { return Selector_1.SelectorClass; } });
exports.Objective = init_1.dataPack.Objective, exports.Selector = init_1.dataPack.Selector, exports.Data = init_1.dataPack.Data, exports.Variable = init_1.dataPack.Variable;
// TYPES
__exportStar(require("./arguments/index"), exports);
__exportStar(require("./datapack/BasePath"), exports);
__exportStar(require("./datapack/Datapack"), exports);
